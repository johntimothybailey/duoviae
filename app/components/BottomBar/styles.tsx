import { Spacing } from '../../theme'
import { FlexStyle, ViewStyle } from 'react-native'

const Main: FlexStyle = {
  flex: 0,
  flexDirection: 'column'
}

const Items: ViewStyle = {
  paddingTop: Spacing.SMALL,
  paddingBottom: Spacing.LARGE
}

export default {
  Main,
  Items
}
