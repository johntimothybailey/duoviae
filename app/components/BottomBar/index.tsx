import React from 'react'
import {
  BottomNavigation,
  BottomNavigationTab,
  BottomNavigationTabElement,
  Divider, Layout
} from '@ui-kitten/components'
import { Route } from '@react-navigation/core'
import { Props } from './props'
import styles from './styles'

/**
 * @cite https://reactnavigation.org/blog/2019/11/04/using-react-navigation-5-with-ui-kitten/#step-3-bottom-tabs
 * @cite https://reactnavigation.org/blog/2019/11/04/using-react-navigation-5-with-ui-kitten/#typescript
 * @param props
 * @constructor
 */
export const BottomBar = (props: Props): React.ReactElement => {
  const onSelect = (index: number): void => {
    const selectedTabRoute: string = props.state.routeNames[index]
    props.navigation.navigate(selectedTabRoute)
  }

  /**
   * Consider moving this out
   * @param route
   */
  const createNavigationTabForRoute = (route: Route<any, any>): BottomNavigationTabElement => {
    const { options } = props.descriptors[route.key] || { options: {} }
    return (
      <BottomNavigationTab
        key={route.key}
        title={options.title ?? route.name}
        // @ts-expect-error: all Tab Screens strictly have UI Kitten Icon
        icon={options.tabBarIcon}
      />
    )
  }

  return (
    <Layout style={styles.Main} level='1'>
      <Divider />
      <BottomNavigation
        {...props}
        style={styles.Items}
        appearance='noIndicator'
        selectedIndex={props.state.index}
        onSelect={onSelect}
      >
        {props.state.routes.map(createNavigationTabForRoute)}
      </BottomNavigation>
    </Layout>
  )
}

export default BottomBar
