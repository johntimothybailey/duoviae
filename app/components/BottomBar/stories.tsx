import * as React from 'react'
import { storiesOf } from '@storybook/react-native'
import BottomBar from './index'
import { action } from '@storybook/addon-actions'
import { Spacing } from '../../theme'
import { Layout } from '@ui-kitten/components'
import { stateItem, stateItems } from './index.test'

declare let module: NodeModule

const onNavigate = (value: any) => action('navigate to')(value)

storiesOf('BottomBar', module)
  .add('One item with no icon nor descriptors', () => (
    <BottomBar state={stateItem} descriptors={{}} navigation={{ navigate: onNavigate } as any} />
  ))
  .add('Items with icons and descriptors', () => (
    <Layout level='4' style={{ paddingHorizontal: Spacing.SMALLER }}>
      <BottomBar
        state={stateItems} descriptors={{
          home: {
            options: {
              title: 'Welcome'
            }
          } as any
        }} navigation={{ navigate: onNavigate } as any}
      />
    </Layout>
  ))
