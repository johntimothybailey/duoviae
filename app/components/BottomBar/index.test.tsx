import React from 'react'
import { cleanup, render, fireEvent } from '@testing-library/react-native'
import { BootComponent } from '../../../test'
import BottomBar from './index'
import { TabNavigationState } from '@react-navigation/native'
import { ParamListBase } from '@react-navigation/routers'

export const stateItem: TabNavigationState<ParamListBase> = {
  key: 'home',
  type: 'tab',
  index: 0,
  routeNames: ['home'],
  routes: [
    {
      key: 'home',
      name: 'Home'
    }
  ],
  stale: false,
  history: []
}

export const stateItems: TabNavigationState<ParamListBase> = {
  key: 'home',
  type: 'tab',
  index: 0,
  routeNames: ['home', 'account'],
  routes: [
    {
      key: 'home',
      name: 'Home'
    },
    {
      key: 'account',
      name: 'Account'
    }
  ],
  stale: false,
  history: []
}

describe('BottomBar', () => {
  describe('Render', () => {
    afterEach(cleanup)
    it('items without descriptors', () => {
      const { queryByText, toJSON } = render(
        <BootComponent>
          <BottomBar
            state={stateItems} descriptors={{}} navigation={{ navigate: () => {} } as any}
          />
        </BootComponent>
      )
      expect(queryByText('Welcome')).not.toBeTruthy()
      expect(queryByText('Home')).toBeTruthy()
      expect(queryByText('Account')).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })
    it('items using descriptors', () => {
      const { queryByText, toJSON } = render(
        <BootComponent>
          <BottomBar
            state={stateItems} descriptors={{
              home: {
                options: {
                  title: 'Welcome'
                }
              } as any
            }} navigation={{ navigate: () => {} } as any}
          />
        </BootComponent>
      )
      expect(queryByText('Welcome')).toBeTruthy()
      expect(queryByText('Home')).not.toBeTruthy()
      expect(queryByText('Account')).toBeTruthy()
      expect(toJSON()).toMatchSnapshot()
    })
    it('triggers navigate on click', () => {
      const onNavigate = jest.fn()
      const { queryByText } = render(
        <BootComponent>
          <BottomBar
            state={stateItems} descriptors={{}} navigation={{ navigate: onNavigate } as any}
          />
        </BootComponent>
      )
      fireEvent.press(queryByText('Account'))
      expect(onNavigate).toHaveBeenCalledWith('account')
    })
  })
})
