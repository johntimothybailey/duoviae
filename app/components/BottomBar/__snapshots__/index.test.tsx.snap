// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomBar Render items using descriptors 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      appearance="default"
      level="1"
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
          },
          Object {
            "flex": 0,
            "flexDirection": "column",
          },
        ]
      }
    >
      <View
        appearance="default"
        style={
          Array [
            Object {
              "backgroundColor": "#EDF1F7",
              "height": 1,
            },
            undefined,
          ]
        }
      />
      <View
        appearance="noIndicator"
        descriptors={
          Object {
            "home": Object {
              "options": Object {
                "title": "Welcome",
              },
            },
          }
        }
        navigation={
          Object {
            "navigate": [Function],
          }
        }
        onSelect={[Function]}
        selectedIndex={0}
        state={
          Object {
            "history": Array [],
            "index": 0,
            "key": "home",
            "routeNames": Array [
              "home",
              "account",
            ],
            "routes": Array [
              Object {
                "key": "home",
                "name": "Home",
              },
              Object {
                "key": "account",
                "name": "Account",
              },
            ],
            "stale": false,
            "type": "tab",
          }
        }
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "minHeight": 56,
              "paddingVertical": 4,
            },
            Object {
              "flexDirection": "row",
            },
            Object {
              "paddingBottom": 24,
              "paddingTop": 12,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            style={
              Array [
                Object {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#3366FF",
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "600",
                  "marginVertical": 2,
                },
              ]
            }
          >
            Welcome
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            style={
              Array [
                Object {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#8F9BB3",
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "600",
                  "marginVertical": 2,
                },
              ]
            }
          >
            Account
          </Text>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;

exports[`BottomBar Render items without descriptors 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      appearance="default"
      level="1"
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
          },
          Object {
            "flex": 0,
            "flexDirection": "column",
          },
        ]
      }
    >
      <View
        appearance="default"
        style={
          Array [
            Object {
              "backgroundColor": "#EDF1F7",
              "height": 1,
            },
            undefined,
          ]
        }
      />
      <View
        appearance="noIndicator"
        descriptors={Object {}}
        navigation={
          Object {
            "navigate": [Function],
          }
        }
        onSelect={[Function]}
        selectedIndex={0}
        state={
          Object {
            "history": Array [],
            "index": 0,
            "key": "home",
            "routeNames": Array [
              "home",
              "account",
            ],
            "routes": Array [
              Object {
                "key": "home",
                "name": "Home",
              },
              Object {
                "key": "account",
                "name": "Account",
              },
            ],
            "stale": false,
            "type": "tab",
          }
        }
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "minHeight": 56,
              "paddingVertical": 4,
            },
            Object {
              "flexDirection": "row",
            },
            Object {
              "paddingBottom": 24,
              "paddingTop": 12,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            style={
              Array [
                Object {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#3366FF",
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "600",
                  "marginVertical": 2,
                },
              ]
            }
          >
            Home
          </Text>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            style={
              Array [
                Object {
                  "color": "#222B45",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                Object {
                  "color": "#8F9BB3",
                  "fontFamily": "System",
                  "fontSize": 12,
                  "fontWeight": "600",
                  "marginVertical": 2,
                },
              ]
            }
          >
            Account
          </Text>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
